crates/topos-p2p/src/behaviour/gossip.rs

```diff
@@ -10,37 +11,32 @@ use libp2p::{
     identity::Keypair,
     swarm::{NetworkBehaviour, THandlerInEvent, ToSwarm},
 };
-use serde::{Deserialize, Serialize};
-use topos_metrics::{
-    P2P_DUPLICATE_MESSAGE_ID_RECEIVED_TOTAL, P2P_GOSSIP_BATCH_SIZE,
-    P2P_MESSAGE_SERIALIZE_FAILURE_TOTAL,
-};
+use prost::Message as ProstMessage;
+use topos_api::grpc::tce::v1::{Batch, DoubleEchoRequest};
+use topos_metrics::{P2P_DUPLICATE_MESSAGE_ID_RECEIVED_TOTAL, P2P_GOSSIP_BATCH_SIZE};
 use tracing::{debug, error};
 
 use crate::{constants, event::ComposedEvent, TOPOS_ECHO, TOPOS_GOSSIP, TOPOS_READY};
 
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub(crate) struct Batch {
-    pub(crate) data: Vec<Vec<u8>>,
-}
-
 pub struct Behaviour {
     batch_size: usize,
     gossipsub: gossipsub::Behaviour,
-    echo_queue: VecDeque<Vec<u8>>,
-    ready_queue: VecDeque<Vec<u8>>,
+    pending: HashMap<&'static str, VecDeque<Vec<u8>>>,
     tick: tokio::time::Interval,
     cache: HashSet<MessageId>,
 }
 
 impl Behaviour {
-    pub fn publish(&mut self, topic: &'static str, data: Vec<u8>) -> Result<usize, &'static str> {
+    pub fn publish(
+        &mut self,
+        topic: &'static str,
+        message: Vec<u8>,
+    ) -> Result<usize, &'static str> {
         match topic {
             TOPOS_GOSSIP => {
-                _ = self.gossipsub.publish(IdentTopic::new(topic), data);
+                _ = self.gossipsub.publish(IdentTopic::new(topic), message);
```

Comment:
I'm wondering if we could not just `Batch` the `gossip` too 👀  what do you think?
Let's create a ticket for that as it is not really linked to this issue.